keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,694 @@
+```
+  --cluster-cidr  ipNetSlice   (IP CIDRs, in a comma separated list, Default: [])
+```
+Only the first CIDR for each IP family will be used; all others will be ignored.
+
+The cloud_cidr_allocator will be updated to support allocating from multiple CIDRs. The route_controller will be updated to create routes for multiple CIDRs.
+
+### Container Environment Variables
+
+The [container environmental variables](https://kubernetes.io/docs/concepts/containers/container-environment-variables/#container-environment) should support dual stack.
+
+Pod information is exposed through environmental variables on the pod. There are a few environmental variables that are automatically created, and some need to be specified in the pod definition, through the downward api.
+
+The Downward API [status.podIP](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#capabilities-of-the-downward-api) will preserve the existing single IP address, and will be set to the default IP for each pod. A new environmental variable named status.podIPs will contain a space-separated list of IP addresses. The new pod API will have a slice of structures for the additional IP addresses. Kubelet will translate the pod structures and return podIPs as a comma-delimited string.
+
+Here is an example of how to define a pluralized MY_POD_IPS environmental variable in a pod definition yaml file:
+```
+  - name: MY_POD_IPS
+    valueFrom:
+      fieldRef:
+        fieldPath: status.podIPs
+  - name: MY_POD_IPS_PROPERTIES_0
+    valueFrom:
+      fieldRef:
+        fieldPath: status.podIPs[0].properties
+```
+
+This definition will cause an environmental variable setting in the pod similar to the following:
+```
+MY_POD_IPS=fd00:10:20:0:3::3,10.20.3.3
+MY_POD_IPS_PROPERTIES_<IP_index>=x:y #We assume that colon is not going to used in map values.
+```
+
+### Kubeadm Support
+
+Dual-stack support will need to be added to kubeadm both for dual-stack development purposes, and for use in dual-stack continuous integration tests.
+
+- The Kubeadm config options and config file will support dual stack options for apiserver-advertise-address, and podSubnet.
+
+#### Kubeadm Configuration Options
+
+The kubeadm configuration options for advertiseAddress and podSubnet will need to be changed to handle a comma-separated list of CIDRs:
+```
+    api:
+      advertiseAddress: "fd00:90::2,10.90.0.2" [Multiple IP CIDRs, comma separated list of CIDRs]
+    networking:
+      podSubnet: "fd00:10:20::/72,10.20.0.0/16" [Multiple IP CIDRs, comma separated list of CIDRs]
+```
+
+#### Kubeadm-Generated Manifests
+
+Kubeadm will need to generate dual-stack CIDRs for the --service-cluster-ip-range command line argument in kube-apiserver.yaml:
```

Comment:
Since the current goal is have single-family services, I don't think we need to support a list of CIDRs in `--service-cluster-ip-range`
Further down in this KEP, the variation of Dual-Stack Service CIDRs is discussed but that's a non-goal. There this change would be applicable and has already been noted.